src/webview/generateInboundEventEditSequence.js

```diff
@@ -0,0 +1,133 @@
+/* @flow strict-local */
+import invariant from 'invariant';
+
+import type { MessageListElement, Narrow, GetText } from '../types';
+import type { BackgroundData } from './MessageList';
+import messageListElementHtml from './html/messageListElementHtml';
+
+export type Insert = {|
+  type: 'insert',
+  html: string,
+  index: number,
+|};
+
+export type Delete = {|
+  type: 'delete',
+  index: number,
+|};
+
+export type Replace = {|
+  type: 'replace',
+  html: string,
+  index: number,
+|};
+
+type Edit = Insert | Delete | Replace;
+
+export type EditSequence = $ReadOnlyArray<Edit>;
+
+/**
+ * A compare function to sort a list of MessageListElements by `.key`.
+ *
+ * See the `MessageListElement.key` type definition.
+ */
+function compare(a: MessageListElement, b: MessageListElement) {
+  const firstResult = Math.sign(a.key[0] - b.key[0]);
+  if (firstResult !== 0) {
+    return firstResult;
+  } else {
+    return Math.sign(a.key[1] - b.key[1]);
+  }
+}
+
+function areElementsInOrder(elements: $ReadOnlyArray<MessageListElement>): boolean {
+  return elements.every((elem, i, arr) => i === 0 || compare(arr[i - 1], arr[i]) === -1);
+}
+
+/**
+ * PRIVATE - don't use until we have more test coverage (coming soon)
+ */
+export function getEditSequence(
+  oldArgs: {|
+    backgroundData: BackgroundData,
+    narrow: Narrow,
+    elements: $ReadOnlyArray<MessageListElement>,
+    _: GetText,
+  |},
+  newArgs: {|
+    backgroundData: BackgroundData,
+    narrow: Narrow,
+    elements: $ReadOnlyArray<MessageListElement>,
+    _: GetText,
+  |},
+): EditSequence {
+  const { elements: oldElements } = oldArgs;
+  const { elements: newElements } = newArgs;
+
+  // TODO: Expensive!
+  invariant(areElementsInOrder(oldElements), 'getEditSequence: oldElements not in order');
+  invariant(areElementsInOrder(newElements), 'getEditSequence: newElements not in order');
+
+  const result = [];
+
+  let i = 0;
+  let j = 0;
+  while (i < oldElements.length && j < newElements.length) {
+    const oldElement = oldElements[i];
+    const newElement = newElements[j];
+    const compareResult = compare(oldElement, newElement);
+    if (compareResult < 0) {
+      result.push({ type: 'delete', index: j });
+      i++;
+    } else if (compareResult > 0) {
+      result.push({
+        type: 'insert',
+        index: j,
+        html: messageListElementHtml({
+          backgroundData: newArgs.backgroundData,
+          narrow: newArgs.narrow,
+          element: newElement,
+          _: newArgs._,
+        }),
+      });
+      j++;
+    } else {
+      const oldHtml = messageListElementHtml({
```

Comment:
There's a performance regression embedded here: we now render each message-list element twice on each update elsewhere in the list, instead of just once.

The existing code takes a shortcut here by assuming that if the element is equal, and all flags except read/unread are equal, then nothing changed:
```js
  if (
    !isEqual(
      prevProps.messageListElementsForShownMessages,
      nextProps.messageListElementsForShownMessages,
    )
    || !equalFlagsExcludingRead(prevProps.backgroundData.flags, nextProps.backgroundData.flags)
  ) {
    uevents.push(updateContent(prevProps, nextProps));
```

I think we can take the same shortcut. I'm pretty sure we'll want *some* kind of shortcut to avoid rerendering untouched messages.

It's fine for it to be this way at this point in the branch, though, before we start actually using this system; just should have a TODO comment.

Reply:
I've made a stab at a shortcut. I think it's a bit better than the above:

- It triggers a `replace` when the language (a.k.a. `_`) changes, in the spirit of live-updating the UI when its source data changes, always. (IOW, it doesn't matter that I don't think you can currently change the language while viewing a message list.) We translate muted-user placeholders in the message list.
- Instead of checking for flags changing on any of the message we know about, it just checks flags on the relevant message, and only when it's a `type: 'message'` element, since that's the only place where we render flags.

Reply:
Cool, yeah, those do seem like improvements.